### PR TITLE
config refactor stage 1 - ExperimentConfig and LocalConfig

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

Port `ExperimentConfig` to the new base class.

Now the canonicalize and validation logic is shorter, and should be far more readable.

### Checklist ###
  - [x] test case
  - [x] doc

### How to test ###

Should be tested together with other configure refactor PRs.

### CODE REVIEW

File: nni/experiment/config/exp_config.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        super()._validate_canonical()
+
+        space_cnt = (self.search_space is not None) + (self.search_space_file is not None)
+        if self.use_annotation and space_cnt != 0:
+            raise ValueError('ExperimentConfig: search space must not be set when annotation is enabled')
+        if not self.use_annotation and space_cnt != 1:
+            raise ValueError('ExperimentConfig: search_space and search_space_file must be set one')
+
+        if self.search_space_file is not None:
+            self.search_space = yaml.safe_load(open(self.search_space_file))
```
Comment (SparkSnail) [inline nni/experiment/config/exp_config.py:116]: does the file need close()?

File: nni/experiment/config/utils/internal.py [hunk 2]
```diff
@@ -108,17 +112,44 @@ def guess_list_config_type(objs, type_hint, _hint_list_item=False):
...
     subclasses = set(cls.__subclasses__())
     return subclasses.union(*[_all_subclasses(subclass) for subclass in subclasses])
+
+## training service factory ##
+
+def training_service_config_factory(platform):
+    from ..training_service import TrainingServiceConfig  # avoid circular import
+
+    # import all custom config classes so they can be found in TrainingServiceConfig.__subclasses__()
+    custom_ts_config_path = nni.runtime.config.get_config_file('training_services.json')
```
Comment (SparkSnail) [inline nni/experiment/config/utils/internal.py:146]: What is content of `training_service.json` file? Does the file has path validation?

File: nni/experiment/config/training_services/local.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    platform: str = 'local'
+    use_active_gpu: Optional[bool] = None
+    max_trial_number_per_gpu: int = 1
+    gpu_indices: Union[List[int], int, str, None] = None
+    reuse_mode: bool = False
+
+    def _canonicalize(self, parents):
+        super()._canonicalize(parents)
+        self.gpu_indices = canonical_gpu_indices(self.gpu_indices)
+        self.nni_manager_ip = None
```
Comment (J-shang) [inline nni/experiment/config/training_services/local.py:36]: do we have `nni_manager_ip` in the training service config before?

File: nni/experiment/config/exp_config.py (hunk 1 above)
Comment (liuzhe-lz) [reply nni/experiment/config/exp_config.py:116]: Well I'm pretty sure it won't cause problem to open a file read-only and let GC to close it, but you are right it's better to use `with`.

File: nni/experiment/config/utils/internal.py (hunk 2 above)
Comment (liuzhe-lz) [reply nni/experiment/config/utils/internal.py:146]: This is a config file for custom training services, similar to the config file for registered algorithms.
The availability is guaranteed by the `config` module.

File: nni/experiment/config/training_services/local.py (hunk 3 above)
Comment (liuzhe-lz) [reply nni/experiment/config/training_services/local.py:36]: `nni_manager_ip`, `trial_command`, `trial_gpu_number`, etc were only in the top level config and not in training service configs.
But it was decided several releases ago, according to discussion, that training service config should contain all things needed by training service. And then these fields in top level config will work as "shortcuts" for convenience and for compatibility.

**Review by J-shang — APPROVED** [review]

**Review by SparkSnail — APPROVED** [review]

**Pull request merged**